### PR TITLE
Store relative file paths for fuzzy finder in sets

### LINKED ISSUE #1418: Duplicated items in "Find In Project Popover"
### What Happened?

Opened popover with a project open and started entering a query. Some of the results I got were duplicates.

### Steps to Reproduce

Unable to reproduce, happens randomly.

Screenshot from 2024-01-25 19 16 31

### Expected Behavior

No duplicated items

### OS Version

7.x (Horus)

### Software Version

Compiled from git

### Log Output

_No response_

### Hardware Info

_No response_

### PR DESCRIPTION
**Pull request by colinkiama**
Fixes #1418

### CODE REVIEW

File: plugins/fuzzy-search/fuzzy-finder.vala [hunk 1]
```diff
@@ -241,60 +241,61 @@ public class Scratch.Services.FuzzyFinder {
...
-                            : 0);
-
-                    results.add (path_search_result);
-                }
-            }
+            Gee.Iterator<string> iter = project.relative_file_paths.iterator ();
+
+            while (iter.next () && !cancellable.is_cancelled ()) {
+              if (cancellable.is_cancelled ()) {
+                  return results;
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-finder.vala:248]: Never reached because of `!cancellable.is_cancelled ()` in previous line?

Comment (colinkiama) [reply plugins/fuzzy-search/fuzzy-finder.vala]: Thanks for catching that! Should be fixed now

**Review by jeremypw — CHANGES_REQUESTED** [review]
Everything seems to work as before except no duplicates could be created - good work. Just an inline comment.  Also there are several places where `var` should be used as the variable type is implicit.

Some searches do not work as desired but the same in master - we need to iterate on the logic some more to get searched like `code/fuzfind` to work as expected from the comment in the code (in a future PR).

**Comment by colinkiama** [comment]
Sorry, I'm not sure what needs to be changed. Is it that I should use `var` instead of `Gee.Iterator` explicitly?

File: plugins/fuzzy-search/fuzzy-finder.vala [hunk 2]
```diff
@@ -241,60 +241,61 @@ public class Scratch.Services.FuzzyFinder {
...
-                            ? CURRENT_PROJECT_PRIORITY_BONUS
-                            : 0);
-
-                    results.add (path_search_result);
-                }
-            }
+            Gee.Iterator<string> iter = project.relative_file_paths.iterator ();
+
+            while (iter.next () && !cancellable.is_cancelled ()) {
+              string path = iter.get ();
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-finder.vala:247]: Use var

File: plugins/fuzzy-search/fuzzy-finder.vala [hunk 3]
```diff
@@ -241,60 +241,61 @@ public class Scratch.Services.FuzzyFinder {
...
+              SearchResult path_search_result;
+              SearchResult filename_search_result;
+
+              // If there is more than one project prepend the project name
+              // to the front of the path
+              // This helps to search for specific files only in one project, e.g.
+              // "code/fuzfind" will probably only return fuzzy_finder.vala from this project
+              // even if their is a "fuzzy_finder" file in another project
+              string project_name = "";
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-finder.vala:256]: Use var.

File: plugins/fuzzy-search/fuzzy-finder.vala [hunk 4]
```diff
@@ -241,60 +241,61 @@ public class Scratch.Services.FuzzyFinder {
...
+
+              project_name = project_paths.size > 1 ? Path.get_basename (project.root_path) : "";
+              if (dir_length > 0 || project_paths.size == 1) {
+                  path_search_result = fuzzy_match (search_str, dir_length, path, cancellable);
+              } else {
+                  path_search_result = fuzzy_match (search_str, dir_length, @"$project_name/$path", cancellable);
+              }
+
+              if (dir_length == 0) {
+                  string filename = Path.get_basename (path);
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-finder.vala:266]: Use var

File: plugins/fuzzy-search/fuzzy-search-project.vala [hunk 5]
```diff
@@ -44,14 +45,15 @@ public class Scratch.Services.SearchProject {
         }
 
         int start_length = relative_file_paths.size;
+        Gee.Iterator<string> iter = relative_file_paths.iterator ();
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-search-project.vala:48]: Use var.

File: plugins/fuzzy-search/fuzzy-search-project.vala [hunk 6]
```diff
@@ -44,14 +45,15 @@ public class Scratch.Services.SearchProject {
...
+
         // Remove directory
-        for (int i = start_length - 1; i > -1; i--) {
-            string relative_path = relative_file_paths[i];
-            if (relative_path.has_prefix (deleted_path)) {
-                relative_file_paths.remove (relative_path);
-            }
+        while (iter.next ()) {
+          string relative_path = iter.get ();
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-search-project.vala:52]: Use var.

File: plugins/fuzzy-search/fuzzy-finder.vala [hunk 7]
```diff
@@ -241,60 +241,61 @@ public class Scratch.Services.FuzzyFinder {
...
-                    path_search_result.project = project_name;
-                    path_search_result.score = (int) (path_search_result.score * 0.2) +
-                        (project.root_path == current_project
-                            ? CURRENT_PROJECT_PRIORITY_BONUS
-                            : 0);
-
-                    results.add (path_search_result);
-                }
-            }
+            Gee.Iterator<string> iter = project.relative_file_paths.iterator ();
```
Comment (jeremypw) [inline plugins/fuzzy-search/fuzzy-finder.vala:244]: Use var.

**Review by jeremypw — CHANGES_REQUESTED** [review]
I have marked where I think `var` can be used.  These are where the type can be inferred from the syntax or the `.vapi`. Usually when there is an assignment as well as a declaration.  If anything doesnt compile just revert to explicit.

File: plugins/fuzzy-search/fuzzy-finder.vala (hunk 7 above)
Comment (colinkiama) [reply plugins/fuzzy-search/fuzzy-finder.vala]: Updated all of these lines now

**Review by jeremypw — APPROVED** [review]

**Pull request merged**